File: mfga-xposed/app/src/main/java/com/mfga/xposed/diagnostics/BadgeDrawObserver.java
package com.mfga.xposed.diagnostics;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.graphics.fonts.Font;
import android.graphics.fonts.FontVariationAxis;
import android.graphics.text.PositionedGlyphs;
import android.graphics.text.TextRunShaper;
import android.util.Log;

import java.io.File;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.BiConsumer;

import io.github.libxposed.api.XposedInterface;

/** Read-only, bounded observation of actual SystemUI draw calls. Never changes arguments. */
public final class BadgeDrawObserver {
    private final BadgeSamplePolicy samples = new BadgeSamplePolicy(12);
    private final ThreadLocal<Boolean> active = new ThreadLocal<>();
    private final AtomicBoolean started = new AtomicBoolean();
    private final AtomicBoolean failed = new AtomicBoolean();
    private final BiConsumer<Integer, String> report;

    public BadgeDrawObserver(BiConsumer<Integer, String> report) { this.report = report; }

    public void install(BiConsumer<Method, XposedInterface.Hooker> installer) {
        if (!started.compareAndSet(false, true)) return;
        int candidates = 0;
        for (String name : new String[]{"android.graphics.Canvas", "android.graphics.BaseCanvas",
                "android.graphics.RecordingCanvas", "android.graphics.BaseRecordingCanvas"}) {
            try {
                Class<?> type = Class.forName(name, false, Canvas.class.getClassLoader());
                for (Method method : type.getDeclaredMethods()) {
                    if (Modifier.isStatic(method.getModifiers()) || method.getReturnType() != void.class) continue;
                    String operation = method.getName();
                    if (!operation.equals("drawText") && !operation.equals("drawTextRun")) continue;
                    Class<?>[] params = method.getParameterTypes();
                    if (params.length < 4 || params[params.length - 1] != Paint.class) continue;
                    if (params[0] != String.class && params[0] != CharSequence.class && params[0] != char[].class) continue;
                    candidates++;
                    installer.accept(method, chain -> {
                        if (samples.full() || Boolean.TRUE.equals(active.get())) return chain.proceed();
                        active.set(true);
                        try {
                            try {
                                capture(method, chain.getThisObject(), chain.getArgs());
                            } catch (RuntimeException | LinkageError error) {
                                if (failed.compareAndSet(false, true)) {
                                    report.accept(Log.WARN, "[badge-observe-failed] drawing unchanged: " + error);
                                }
                            }
                            // Original drawing and its exceptions always proceed untouched.
                            return chain.proceed();
                        } finally {
                            active.remove();
                        }
                    });
                }
            } catch (ClassNotFoundException | LinkageError absent) {
                // Different Android builds may expose different canvas implementation classes.
            }
        }
        report.accept(Log.INFO, "[badge-observe-ready] read-only; samples=7,10; limit=12; candidate_methods=" + candidates);
    }

    private void capture(Method method, Object receiver, List<Object> args) {
        if (!(receiver instanceof Canvas)) return;
        Object text = args.get(0);
        if (text == null) return;
        int start = 0;
        int end;
        int xIndex;
        boolean rtl = false;
        if (args.size() == 4 && method.getName().equals("drawText")) {
            end = ((CharSequence) text).length();
            xIndex = 1;
        } else if (args.size() == 6 && method.getName().equals("drawText")) {
            start = (Integer) args.get(1);
            int extent = (Integer) args.get(2);
            end = text instanceof char[] ? start + extent : extent;
            xIndex = 3;
        } else if (args.size() == 9 && method.getName().equals("drawTextRun")) {
            start = (Integer) args.get(1);
            int extent = (Integer) args.get(2);
            end = text instanceof char[] ? start + extent : extent;
            xIndex = 5;
            rtl = (Boolean) args.get(7);
        } else {
            return;
        }
        String sample = BadgeSamplePolicy.sample(text, start, end);
        if (sample == null) return;
        Paint original = (Paint) args.get(args.size() - 1);
        if (original == null || original.getTextSize() <= 0 || original.getTextSize() > 96) return;
        String caller = callers();
        Typeface face = original.getTypeface();
        String key = caller + '|' + sample + '|' + original.getTextSize() + '|'
                + (face == null ? 0 : face.hashCode());
        if (!samples.claim(key)) return;

        // Measuring a copy cannot change the Paint supplied to the original draw call.
        Paint paint = new Paint(original);
        float x = (Float) args.get(xIndex);
        float y = (Float) args.get(xIndex + 1);
        Rect clip = new Rect();
        ((Canvas) receiver).getClipBounds(clip);
        Paint.FontMetrics fm = paint.getFontMetrics();
        Paint.FontMetricsInt fmi = paint.getFontMetricsInt();
        Rect textBounds = new Rect();
        paint.getTextBounds(sample, 0, sample.length(), textBounds);
        PositionedGlyphs run = TextRunShaper.shapeTextRun(sample, 0, sample.length(), 0,
                sample.length(), 0, 0, rtl, paint);
        RectF ink = new RectF();
        boolean hasInk = false;
        Set<String> fonts = new LinkedHashSet<>();
        for (int i = 0; i < run.glyphCount(); i++) {
            Font font = run.getFont(i);
            File file = font.getFile();
            StringBuilder identity = new StringBuilder(file == null ? "<memory-font>" : file.getPath());
            identity.append(" weight=").append(font.getStyle().getWeight());
            FontVariationAxis[] axes = font.getAxes();
            if (axes != null) for (FontVariationAxis axis : axes) {
                identity.append(' ').append(axis.getTag()).append('=').append(axis.getStyleValue());
            }
            fonts.add(identity.toString());
            RectF bounds = new RectF();
            font.getGlyphBounds(run.getGlyphId(i), paint, bounds);
            bounds.offset(run.getGlyphX(i), run.getGlyphY(i));
            if (!bounds.isEmpty()) {
                if (hasInk) ink.union(bounds); else ink.set(bounds);
                hasInk = true;
            }
        }
        report.accept(Log.INFO, "[badge-sample] text=" + sample + " px=" + paint.getTextSize()
                + " x=" + x + " baseline=" + y + " clip=" + clip.toShortString()
                + " canvas=" + receiver.getClass().getName() + " method=" + method.getName());
        report.accept(Log.INFO, String.format(Locale.ROOT,
                "[badge-metrics] float=%s int=%s run=(%.3f,%.3f) textBounds=%s ink=%s typefaceWeight=%s align=%s",
                fm.top + "," + fm.ascent + "," + fm.descent + "," + fm.bottom,
                fmi.top + "," + fmi.ascent + "," + fmi.descent + "," + fmi.bottom,
                run.getAscent(), run.getDescent(), textBounds.toShortString(),
                hasInk ? ink.toShortString() : "EMPTY", face == null ? "default" : face.getWeight(), paint.getTextAlign()));
        for (String font : fonts) report.accept(Log.INFO, "[badge-font] " + font);
        report.accept(Log.INFO, "[badge-caller] " + caller);
        if (samples.full()) report.accept(Log.INFO, "[badge-observe-limit] sampling stopped; all drawing still unchanged");
    }

    private static String callers() {
        StringBuilder result = new StringBuilder();
        int count = 0;
        for (StackTraceElement frame : Thread.currentThread().getStackTrace()) {
            String name = frame.getClassName();
            if (name.startsWith("com.mfga.xposed.") || name.startsWith("io.github.libxposed.")
                    || name.startsWith("org.lsposed.") || name.startsWith("de.robv.android.xposed.")
                    || name.startsWith("java.") || name.startsWith("android.graphics.")) continue;
            if (result.length() != 0) result.append(" <- ");
            result.append(name).append('.').append(frame.getMethodName());
            if (++count == 10) break;
        }
        return result.toString();
    }
}

File: mfga-xposed/app/src/main/kotlin/com/mfga/xposed/diagnostics/BadgeDrawObserver.kt
package com.mfga.xposed.diagnostics

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.RectF
import android.graphics.fonts.Font
import android.graphics.text.TextRunShaper
import android.util.Log
import io.github.libxposed.api.XposedInterface
import java.io.File
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import java.util.Locale
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.BiConsumer

/** Read-only, bounded observation of actual SystemUI draw calls. Never changes arguments. */
class BadgeDrawObserver(private val report: BiConsumer<Int, String>) {
    private val samples = BadgeSamplePolicy(12)
    private val active = ThreadLocal<Boolean>()
    private val started = AtomicBoolean()
    private val failed = AtomicBoolean()

    fun install(installer: BiConsumer<Method, XposedInterface.Hooker>) {
        if (!started.compareAndSet(false, true)) return
        var candidates = 0
        for (name in arrayOf(
            "android.graphics.Canvas", "android.graphics.BaseCanvas",
            "android.graphics.RecordingCanvas", "android.graphics.BaseRecordingCanvas"
        )) {
            try {
                val type = Class.forName(name, false, Canvas::class.java.classLoader)
                for (method in type.declaredMethods) {
                    if (Modifier.isStatic(method.modifiers) || method.returnType != Void.TYPE) continue
                    val operation = method.name
                    if (operation != "drawText" && operation != "drawTextRun") continue
                    val params = method.parameterTypes
                    if (params.size < 4 || params[params.size - 1] != Paint::class.java) continue
                    if (params[0] != String::class.java && params[0] != CharSequence::class.java &&
                        params[0] != CharArray::class.java
                    ) continue
                    candidates++
                    installer.accept(method, XposedInterface.Hooker { chain ->
                        if (samples.full() || active.get() == true) return@Hooker chain.proceed()
                        active.set(true)
                        try {
                            try {
                                capture(method, chain.thisObject, chain.args)
                            } catch (error: Throwable) {
                                when (error) {
                                    is RuntimeException, is LinkageError ->
                                        if (failed.compareAndSet(false, true)) {
                                            report.accept(
                                                Log.WARN,
                                                "[badge-observe-failed] drawing unchanged: $error"
                                            )
                                        }
                                    else -> throw error
                                }
                            }
                            // Original drawing and its exceptions always proceed untouched.
                            chain.proceed()
                        } finally {
                            active.remove()
                        }
                    })
                }
            } catch (absent: Throwable) {
                when (absent) {
                    is ClassNotFoundException, is LinkageError -> {
                        // Different Android builds may expose different canvas implementation classes.
                    }
                    else -> throw absent
                }
            }
        }
        report.accept(
            Log.INFO,
            "[badge-observe-ready] read-only; samples=7,10; limit=12; candidate_methods=$candidates"
        )
    }

    private fun capture(method: Method, receiver: Any?, args: List<Any?>) {
        if (receiver !is Canvas) return
        val text = args[0] ?: return
        var start = 0
        val end: Int
        val xIndex: Int
        var rtl = false
        if (args.size == 4 && method.name == "drawText") {
            end = (text as CharSequence).length
            xIndex = 1
        } else if (args.size == 6 && method.name == "drawText") {
            start = args[1] as Int
            val extent = args[2] as Int
            end = if (text is CharArray) start + extent else extent
            xIndex = 3
        } else if (args.size == 9 && method.name == "drawTextRun") {
            start = args[1] as Int
            val extent = args[2] as Int
            end = if (text is CharArray) start + extent else extent
            xIndex = 5
            rtl = args[7] as Boolean
        } else {
            return
        }
        val sample = BadgeSamplePolicy.sample(text, start, end) ?: return
        val original = args[args.size - 1] as Paint?
        // Badge counts are small text. Skip large draws such as the security keypad (~70px)
        // so the bounded sample budget is spent on the notification badge itself.
        if (original == null || original.textSize <= 0 || original.textSize > 48) return
        val caller = callers()
        // The on-screen number keypad also paints "7"/"10"; it is not a badge. Read-only skip.
        if (caller.contains("eyboard")) return
        val face = original.typeface
        val key = caller + '|' + sample + '|' + original.textSize + '|' +
            (face?.hashCode() ?: 0)
        if (!samples.claim(key)) return

        // Measuring a copy cannot change the Paint supplied to the original draw call.
        val paint = Paint(original)
        val x = args[xIndex] as Float
        val y = args[xIndex + 1] as Float
        val clip = Rect()
        receiver.getClipBounds(clip)
        val fm = paint.fontMetrics
        val fmi = paint.fontMetricsInt
        val textBounds = Rect()
        paint.getTextBounds(sample, 0, sample.length, textBounds)
        val run = TextRunShaper.shapeTextRun(sample, 0, sample.length, 0, sample.length, 0f, 0f, rtl, paint)
        val ink = RectF()
        var hasInk = false
        val fonts = LinkedHashSet<String>()
        for (i in 0 until run.glyphCount()) {
            val font = run.getFont(i)
            val file = font.file
            val identity = StringBuilder(file?.path ?: "<memory-font>")
            identity.append(" weight=").append(font.style.weight)
            val axes = font.axes
            if (axes != null) for (axis in axes) {
                identity.append(' ').append(axis.tag).append('=').append(axis.styleValue)
            }
            fonts.add(identity.toString())
            val bounds = RectF()
            font.getGlyphBounds(run.getGlyphId(i), paint, bounds)
            bounds.offset(run.getGlyphX(i), run.getGlyphY(i))
            if (!bounds.isEmpty) {
                if (hasInk) ink.union(bounds) else ink.set(bounds)
                hasInk = true
            }
        }
        report.accept(
            Log.INFO,
            "[badge-sample] text=$sample px=${paint.textSize}" +
                " x=$x baseline=$y clip=${clip.toShortString()}" +
                " canvas=${receiver.javaClass.name} method=${method.name}"
        )
        report.accept(
            Log.INFO,
            String.format(
                Locale.ROOT,
                "[badge-metrics] float=%s int=%s run=(%.3f,%.3f) textBounds=%s ink=%s typefaceWeight=%s align=%s",
                "${fm.top},${fm.ascent},${fm.descent},${fm.bottom}",
                "${fmi.top},${fmi.ascent},${fmi.descent},${fmi.bottom}",
                run.ascent, run.descent, textBounds.toShortString(),
                if (hasInk) ink.toShortString() else "EMPTY",
                if (face == null) "default" else face.weight, paint.textAlign
            )
        )
        for (font in fonts) report.accept(Log.INFO, "[badge-font] $font")
        report.accept(Log.INFO, "[badge-caller] $caller")
        if (samples.full()) {
            report.accept(Log.INFO, "[badge-observe-limit] sampling stopped; all drawing still unchanged")
        }
    }

    private fun callers(): String {
        val result = StringBuilder()
        var kept = 0
        var appFrames = 0
        for (frame in Thread.currentThread().stackTrace) {
            val name = frame.className
            if (name.startsWith("com.mfga.xposed.") || name.startsWith("io.github.libxposed.") ||
                name.startsWith("org.lsposed.") || name.startsWith("de.robv.android.xposed.") ||
                name.startsWith("java.") || name.startsWith("dalvik.") ||
                name.startsWith("android.graphics.")
            ) continue
            val framework = name.startsWith("android.")
            // Keep the immediate drawing context (first frames), then only non-framework
            // frames up the hierarchy. The badge inherits onDraw from TextView, so its
            // concrete class never shows here; its named SystemUI/Oplus container does.
            if (framework && kept >= 8) continue
            if (!framework) appFrames++
            if (result.isNotEmpty()) result.append(" <- ")
            result.append(name).append('.').append(frame.methodName)
            kept++
            if (appFrames >= 12 || kept >= 40) break
        }
        return result.toString()
    }
}
